opensquirrel/circuit_builder.py

```diff
@@ -31,32 +37,35 @@ class CircuitBuilder(GateLibrary):
     def __init__(
         self,
         number_of_qubits: int,
-        gate_set: [Callable[..., Gate]] = default_gate_set,
+        gate_set: list[Callable[..., Gate]] = default_gate_set,
         gate_aliases: Dict[str, Callable[..., Gate]] = default_gate_aliases,
     ):
         GateLibrary.__init__(self, gate_set, gate_aliases)
         self.squirrel_ir = SquirrelIR(
             number_of_qubits=number_of_qubits, qubit_register_name=self._default_qubit_register_name
         )
 
-    def __getattr__(self, attr):
-        def add_comment(comment_string: str):
-            self.squirrel_ir.add_comment(Comment(comment_string))
-            return self
+    def __getattr__(self, attr: Any) -> Callable[..., Self]:
+        if attr == "comment":
+            return self._add_comment
 
-        def add_this_gate(*args):
-            generator_f = GateLibrary.get_gate_f(self, attr)
+        return partial(self._add_this_gate, attr)
 
-            for i, par in enumerate(inspect.signature(generator_f).parameters.values()):
-                if not isinstance(args[i], par.annotation):
-                    raise TypeError(
-                        f"Wrong argument type for gate `{attr}`, got {type(args[i])} but expected {par.annotation}"
-                    )
+    def to_circuit(self) -> Circuit:
+        return Circuit(self.squirrel_ir)
 
-            self.squirrel_ir.add_gate(generator_f(*args))
-            return self
+    def _add_comment(self, comment_string: str) -> Self:
+        self.squirrel_ir.add_comment(Comment(comment_string))
+        return self
 
-        return add_comment if attr == "comment" else add_this_gate
+    def _add_this_gate(self, attr: Any, *args: Any) -> Self:
+        generator_f = GateLibrary.get_gate_f(self, attr)
```

Comment:
On another note, why is `GateLibrary.get_gate_f(self, attr)` used instead of `self.get_gate_f(attr)`? I see this happening at more places throughout the code, what is the benefit of calling the function in this way?

Reply:
Good spot! I don't see any benefit myself. The reason behind this may be that these functions were once static or class functions. I would open an issue and address this changes in a separate task.